tracer/src/Datadog.Trace/Debugger/Snapshots/DebuggerSnapshotCreator.cs

```diff
@@ -4,31 +4,134 @@
+        public DebuggerSnapshotCreator(bool isFullSnapshot, ProbeLocation location)
         {
+            _isFullSnapshot = isFullSnapshot;
+            _probeLocation = location;
             _jsonUnderlyingString = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
             _jsonWriter = new JsonTextWriter(new StringWriter(_jsonUnderlyingString));
+            _methodScopeMembers = null;
+            CaptureBehaviour = CaptureBehaviour.Capture;
+            _startTime = DateTimeOffset.UtcNow;
+            IsInitialized = false;
+            Initialize();
+        }
+
+        internal bool IsInitialized { get; private set; }
+
+        internal MethodScopeMembers MethodScopeMembers => _methodScopeMembers;
+
+        internal CaptureBehaviour CaptureBehaviour { get; set; }
+
+        public static DebuggerSnapshotCreator BuildSnapshotCreator(string probeId)
+        {
+            var probeInfo = ProbeExpressionsProcessor.Instance.GetProbeInfo(probeId);
+            if (probeInfo.HasValue)
+            {
+                return new DebuggerSnapshotCreator(probeInfo.Value.IsFullSnapshot, probeInfo.Value.ProbeLocation);
+            }
+
+            throw new InvalidOperationException("Probe info not found for probe id: " + probeId);
+        }
+
+        internal CaptureBehaviour DefineSnapshotBehavior<TCapture>(CaptureInfo<TCapture> info, EvaluateAt evaluateAt, bool hasCondition)
+        {
+            if (hasCondition &&
+                evaluateAt == EvaluateAt.Exit &&
+                info.MethodState is MethodState.BeginLine or MethodState.EntryStart)
+            {
+                // in case there is a condition in exit but we are in the entry, don't save entry scope members
+                CaptureBehaviour = CaptureBehaviour.NoCapture;
+                return CaptureBehaviour;
+            }
+
+            if (!hasCondition && _isFullSnapshot)
+            {
+                CaptureBehaviour = CaptureBehaviour.Capture;
+            }
+            else
+            {
+                CaptureBehaviour = CaptureBehaviour.Delayed;
+            }
+
+            if (info.MethodState.IsInAsyncMethod())
+            {
+                CreateMethodScopeMembers(info.AsyncCaptureInfo.HoistedLocals.Length, info.AsyncCaptureInfo.HoistedArguments.Length);
+            }
+            else
+            {
+                CreateMethodScopeMembers(info.LocalsCount.Value, info.ArgumentsCount.Value);
+            }
+
+            return CaptureBehaviour;
+        }
+
+        internal void CreateMethodScopeMembers(int numberOfLocals, int numberOfArguments)
+        {
+            Interlocked.CompareExchange(ref _methodScopeMembers, new MethodScopeMembers(numberOfLocals, numberOfArguments), null);
+        }
+
+        internal void AddScopeMember<T>(string name, Type type, T value, ScopeMemberKind memberKind)
+        {
+            type = type.IsGenericTypeDefinition ? value.GetType() : type;
+            switch (memberKind)
```

Comment:
Is it on purpose that you don't handle `ScopeMemberKind.Argument` & `ScopeMemberKind.Local`?